Xplat/src/main/java/vazkii/botania/data/LooniumStructureLootProvider.java

```diff
@@ -0,0 +1,255 @@
+import net.minecraft.world.level.storage.loot.functions.SetPotionFunction;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.nio.file.Path;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+import static vazkii.botania.common.lib.ResourceLocationHelper.prefix;
+
+public class LooniumStructureLootProvider implements DataProvider {
+	// loot collections based on which village type hoses can actually have chests
+	private static final EnumSet<VillageLoot> PLAINS_VILLAGE_LOOT = EnumSet
+			.of(VillageLoot.CARTOGRAPHER, VillageLoot.FISHER, VillageLoot.TANNERY, VillageLoot.WEAPONSMITH);
+	private static final EnumSet<VillageLoot> DESERT_VILLAGE_LOOT = EnumSet
+			.of(VillageLoot.TEMPLE, VillageLoot.TOOLSMITH, VillageLoot.WEAPONSMITH);
+	private static final EnumSet<VillageLoot> SAVANNA_VILLAGE_LOOT = EnumSet
+			.of(VillageLoot.BUTCHER, VillageLoot.CARTOGRAPHER, VillageLoot.MASON, VillageLoot.TANNERY, VillageLoot.WEAPONSMITH);
+	private static final EnumSet<VillageLoot> SNOWY_VILLAGE_LOOT = EnumSet
+			.of(VillageLoot.ARMORER, VillageLoot.CARTOGRAPHER, VillageLoot.SHEPHERD, VillageLoot.TANNERY, VillageLoot.WEAPONSMITH);
+	private static final EnumSet<VillageLoot> TAIGA_VILLAGE_LOOT = EnumSet
+			.of(VillageLoot.CARTOGRAPHER, VillageLoot.FLETCHER, VillageLoot.TANNERY, VillageLoot.TOOLSMITH, VillageLoot.WEAPONSMITH);
+
+	private final PackOutput.PathProvider pathProvider;
+
+	public LooniumStructureLootProvider(PackOutput packOutput) {
+		this.pathProvider = packOutput.createPathProvider(PackOutput.Target.DATA_PACK, "loot_tables/loonium");
+	}
+
+	private ResourceLocation getLootId(ResourceKey<Structure> structureKey) {
+		return getLootId(structureKey.location());
+	}
+
+	private ResourceLocation getLootId(ResourceLocation structureId) {
+		return prefix("%s/%s".formatted(structureId.getNamespace(), structureId.getPath()));
+	}
+
+	@NotNull
+	@Override
+	public CompletableFuture<?> run(@NotNull CachedOutput cache) {
+		Map<ResourceLocation, LootTable.Builder> tables = new HashMap<>();
+
+		// Note: As far as world generating is concerned, dungeons are "features" (i.e. like trees or geodes),
+		// not "structures" (like everything else the Loonium might care about).
+		tables.put(prefix("default"), buildDelegateLootTable(BuiltInLootTables.SIMPLE_DUNGEON));
+
+		/*
+		Note: Be careful about adding individual items instead of loot table references.
+		The Loonium will randomly either generate a virtual chest full of loot from any referenced table, or put just
+		one of the defined item entries into the "chest". Either way it only picks a single stack from that "chest".
+		Individual item entries need to be weighted accordingly to not have them picked too often.
+		Also, archaeology loot tables need to be handled carefully, due to their limited loot pool options.
+		*/
+
+		tables.put(getLootId(BuiltinStructures.ANCIENT_CITY),
```

Comment:
This could use a helper function, probably. Nicer to author and edit. Maybe an interface like:
```
put(tables, BuiltinStructures.ANCIENT_CITY,
  LootTableReference.lootTableReference(BuiltInLootTables.ANCIENT_CITY).setWeight(4),
  LootTableReference.lootTableReference(BuiltInLootTables.ANCIENT_CITY_ICE_BOX).setWeight(1)
)
```
I could do this myself, sounds like a fun refactor :3

Reply:
Ended up deciding this isn't super necessary